Sources/GRPCCore/Server.swift

```diff
@@ -0,0 +1,428 @@
+  /// listening then all open transports are closed and a ``ServerError`` is thrown.
+  ///
+  /// This function returns when all transports have stopped listening and all requests have been
+  /// handled. You can signal to transports that they should stop listening by calling
+  /// ``stopListening()``. The server will continue to process existing requests.
+  ///
+  /// To stop the server more abruptly you can cancel the task that this function is running in.
+  ///
+  /// You must register all resources you wish to use with the server before calling this function
+  /// as changes made after calling ``run()`` won't be reflected.
+  ///
+  /// - Note: You can only call this function once, repeated calls will result in a
+  ///   ``ServerError`` being thrown.
+  public func run() async throws {
+    try self.state.withLockedValue { state in
+      switch state {
+      case .notStarted:
+        state = .starting
+
+      case .starting, .running:
+        throw ServerError(
+          code: .serverIsAlreadyRunning,
+          message: "The server is already running and can only be started once."
+        )
+
+      case .stopping, .stopped:
+        throw ServerError(
+          code: .serverIsStopped,
+          message: "The server has stopped and can only be started once."
+        )
+      }
+    }
+
+    // When we exit this function we must have stopped.
+    defer {
+      self.state.withLockedValue { $0 = .stopped }
+    }
+
+    var listeners: [RPCAsyncSequence<Stream>] = []
+    listeners.reserveCapacity(self.transports.values.count)
+
+    for transport in self.transports.values {
+      do {
+        let listener = try await transport.listen()
+        listeners.append(listener)
+      } catch let cause {
+        // Failed to start, so start stopping.
+        self.state.withLockedValue { $0 = .stopping }
+        // Some listeners may have started and have streams which need closing.
+        await self.rejectRequests(listeners)
+
+        throw ServerError(
+          code: .failedToStartTransport,
+          message: """
+            Server didn't start because the '\(type(of: transport))' transport threw an error \
+            while starting.
+            """,
+          cause: cause
+        )
+      }
+    }
+
+    // May have been told to stop listening while starting the transports.
+    let isStopping = self.state.withLockedValue { state in
+      switch state {
+      case .notStarted, .running, .stopped:
+        fatalError("Invalid state")
+
+      case .starting:
+        state = .running
+        return false
+
+      case .stopping:
+        return true
+      }
+    }
+
+    // If the server is stopping then notify the transport and then consume them: there may be
+    // streams opened at a lower level (e.g. HTTP/2) which are already open and need to be consumed.
+    if isStopping {
+      for transport in self.transports.values {
+        transport.stopListening()
+      }
```

Comment:
We're already calling `stopListening()` in `rejectRequests(_:)` - we should be able to get rid of this

Reply:
Good catch, this was leftover from a refactor.